### PR TITLE
[Feature] Support MSDeform attn pixel decoder

### PR DESCRIPTION
**Pull request by chhluo**
## Motivation

Take MSDoformAttnPixelDocder out of the Mask2Former #6938 , make it a new small PR.

## Modification

1 add MSDoformAttnPixelDocder which will be used in Mask2Former.
2 add unit test.
3. update format and docstring.

### CODE REVIEW

File: mmdet/models/plugins/msdeformattn_pixel_decoder.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+@PLUGIN_LAYERS.register_module()
+class MSDeformAttnPixelDecoder(BaseModule):
+    """Pixel decoder with multi-scale deformable attention.
+
+    Args:
+        in_channels (list[int] | tuple[int]): Number of channels in the
+            input feature maps.
+        feat_channels (int): Number of channels for feature.
+        out_channels (int): Number of channels for output.
+        num_return_feat_levels (int): Number of levels of feature to
```
Comment (ZwwWayne) [inline mmdet/models/plugins/msdeformattn_pixel_decoder.py:23]: num_return_feat_levels -> num_outs. Number of output feature levels.
(be consistent with FPN)

File: mmdet/models/plugins/msdeformattn_pixel_decoder.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+                self.num_input_levels - self.num_encoder_feat_levels - 1, -1):
+            input_proj = ConvModule(
+                in_channels[i],
+                feat_channels,
+                kernel_size=1,
+                norm_cfg=norm_cfg,
+                act_cfg=None,
+                bias=True)
+            input_proj_list.append(input_proj)
+        self.input_projs = ModuleList(input_proj_list)
```
Comment (ZwwWayne) [inline mmdet/models/plugins/msdeformattn_pixel_decoder.py:89]: self.input_projs -> self.input_convs, input_proj -> input_conv
To be consistent with output_convs, lateral_convs

File: mmdet/models/plugins/msdeformattn_pixel_decoder.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        _, H, W = mask.shape
+        valid_H = torch.sum(~mask[:, :, 0], 1)
+        valid_W = torch.sum(~mask[:, 0, :], 1)
+        valid_ratio_h = valid_H.float() / H
+        valid_ratio_w = valid_W.float() / W
+        valid_ratio = torch.stack([valid_ratio_w, valid_ratio_h], -1)
+        return valid_ratio
+
+    @staticmethod
+    def get_reference_points(spatial_shapes, valid_ratios, device):
```
Comment (ZwwWayne) [inline mmdet/models/plugins/msdeformattn_pixel_decoder.py:169]: We might have function used by Anchor-Free detectors that can be reused here, like that in pointgenerator.

File: mmdet/models/plugins/msdeformattn_pixel_decoder.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+                 positional_encoding=dict(
+                     type='SinePositionalEncoding',
+                     num_feats=128,
+                     normalize=True),
+                 init_cfg=None):
+        super().__init__(init_cfg=init_cfg)
+        self.num_input_levels = len(in_channels)
+        self.num_encoder_feat_levels = \
+            encoder.transformerlayers.attn_cfgs.num_levels
+        assert self.num_encoder_feat_levels >= 1, \
```
Comment (ZwwWayne) [inline mmdet/models/plugins/msdeformattn_pixel_decoder.py:74]: num_encoder_feat_levels -> num_encoder_levels.

File: mmdet/models/plugins/msdeformattn_pixel_decoder.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+                         operation_order=('self_attn', 'norm', 'ffn', 'norm')),
+                     init_cfg=None),
+                 positional_encoding=dict(
+                     type='SinePositionalEncoding',
+                     num_feats=128,
+                     normalize=True),
+                 init_cfg=None):
+        super().__init__(init_cfg=init_cfg)
+        self.num_input_levels = len(in_channels)
+        self.num_encoder_feat_levels = \
```
Comment (ZwwWayne) [inline mmdet/models/plugins/msdeformattn_pixel_decoder.py:72]: self.num_encoder_levels.

File: mmdet/models/plugins/msdeformattn_pixel_decoder.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        # (num_total_query, batch_size, c) -> (batch_size, c, num_total_query)
+        memory = memory.permute(1, 2, 0)
+        c = memory.shape[1]
+
+        # from low resolution to high resolution
+        num_query_per_level = [e[0] * e[1] for e in spatial_shapes]
+        outs = torch.split(memory, num_query_per_level, dim=-1)
+        outs = [
+            x.reshape(batch_size, c, spatial_shapes[i][0],
+                      spatial_shapes[i][1]) for i, x in enumerate(outs)
```
Comment (RangiLyu) [inline mmdet/models/plugins/msdeformattn_pixel_decoder.py:285]: use `x.reshape(batch_size, -1, spatial_shapes[i][0], spatial_shapes[i][1])`
no need to get the channel number

**Review by RangiLyu — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**